### PR TITLE
Kraken, Binance fetchers and aggregation improvements

### PR DESCRIPTION
**Pull request by hatskier**

### CODE REVIEW

File: aggregators/median-aggregator.ts [hunk 1]
```diff
@@ -9,28 +9,53 @@ import {
 const logger =
   require("../utils/logger")("aggregators/median-aggregator") as Consola;
 
-const MAX_DEVIATION = 20; // perecents
+const MAX_DEVIATION = 25; // perecents
```
Comment (jakub-wojciechowski) [inline aggregators/median-aggregator.ts:12]: I'll put this parameter to user manifest. It's also dependent on token type (1-2% for stable coins, 5% for high-liquidity tokens and 25% for the rest). We can create a separate task for it.

File: fetchers/any-swap-fetcher.ts [hunk 2]
```diff
@@ -35,9 +37,23 @@ export default {
...
-        const response = await graphProxy.executeQuery(
+        const fetchStartTime = Date.now();
+        let response = await graphProxy.executeQuery(
           config.subgraphUrl,
           query);
+        const timeElapsed = Date.now() - fetchStartTime;
+
+        // Retrying to fetch if needed
+        if (response.data === undefined && timeElapsed <= RETRY_TIME_LIMIT) {
+          if (timeElapsed <= 2000) {
```
Comment (jakub-wojciechowski) [inline fetchers/any-swap-fetcher.ts]: Hardcoded 2000 -> fix please to RETRY....
And we can move the condition from the if above to reduce duplication

File: fetchers/binance/index.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+const URL = "https://api.binance.com/api/v3/ticker/price";
+
+const binanceFetcher: Fetcher = {
+  async fetchAll(tokenSymbols: string[]): Promise<PriceDataFetched[]> {
+    if (tokenSymbols.length === 0) {
+      logger.warn(`Binance fetcher received an empty array of symbols`);
+      return [];
+    }
+
+    // Fetching ptices from Binance API
```
Comment (jakub-wojciechowski) [inline fetchers/binance/index.ts]: typo

File: fetchers/binance/index.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import axios from "axios";
+import { PriceDataFetched, Fetcher } from "../../types";
+
+const logger =
+  require("../../utils/logger")("fetchers/binance") as Consola;
+
+const URL = "https://api.binance.com/api/v3/ticker/price";
+
+const binanceFetcher: Fetcher = {
+  async fetchAll(tokenSymbols: string[]): Promise<PriceDataFetched[]> {
```
Comment (jakub-wojciechowski) [inline fetchers/binance/index.ts:12]: Hmm, I think we should catch this error a level higher, not to dupliate the code. We can note it as an additional task/refactor with low priority.

File: fetchers/kraken/index.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+const logger =
+  require("../../utils/logger")("fetchers/kraken") as Consola;
+const supportedPairs: string[] = require("./kraken-supported-pairs.json");
+
+const URL = "https://api.kraken.com/0/public/Ticker";
+
+const krakenFetcher: Fetcher = {
+  async fetchAll(tokenSymbols: string[]): Promise<PriceDataFetched[]> {
+    if (tokenSymbols.length === 0) {
```
Comment (jakub-wojciechowski) [inline fetchers/kraken/index.ts]: The same as before - should be moved to caller method to reduce code duplication

File: package.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "limestone-node",
-  "version": "2.0.0",
+  "version": "3.0.0",
```
Comment (jakub-wojciechowski) [inline package.json:3]: We've discussed to change versioning to minor -> 0.3.0

File: runner.ts [hunk 7]
```diff
@@ -306,23 +307,30 @@ export default class Runner {
...
-    return prices.map((p) =>
-      aggregators[this.manifest.priceAggregator].getAggregatedValue(p));
+    const aggregatedPrices: PriceDataAfterAggregation[] = [];
+    const aggregator = aggregators[this.manifest.priceAggregator];
+    for (const price of prices) {
+      try {
+        const priceAfterAggregation = aggregator.getAggregatedValue(price);
+        aggregatedPrices.push(priceAfterAggregation);
+      } catch (e) {
+        logger.error(e.stack);
```
Comment (jakub-wojciechowski) [inline runner.ts:328]: I'm wondering what should we put if there is an error in aggregation, undefined?

File: tools/arweave/check-failed-transactions.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
+const axios = require("axios");
+const Arweave = require("arweave/node");
+
+// const LIMIT = 300;
+const SKIP = 5;
+const LIMIT = 24 * 60;
```
Comment (jakub-wojciechowski) [inline tools/arweave/check-failed-transactions.js:6]: What does the limit stand for - is it the number of recent transactions?

File: utils/arweave-proxy.ts [hunk 9]
```diff
@@ -45,8 +47,15 @@ export default class ArweaveProxy  {
   // This method creates and signs arweave transaction
   // It doesn't post transaction to arweave, to do so use postTransaction
   async prepareUploadTransaction(tags: any, data: any): Promise<Transaction> {
+    const stringifiedData = JSON.stringify(data);
+
+    // Compressing
+    const gzipPromisified = util.promisify(gzip);
+    const gzippedData = await gzipPromisified(stringifiedData);
```
Comment (jakub-wojciechowski) [inline utils/arweave-proxy.ts:54]: How do we handle compression errors - will they block the execution?

**Review by jakub-wojciechowski — CHANGES_REQUESTED** [review]

File: tools/arweave/check-failed-transactions.js (hunk 8 above)
Comment (hatskier) [reply tools/arweave/check-failed-transactions.js:6]: yes, it's the number of recent transactions

**Review by jakub-wojciechowski — APPROVED** [review]

**Pull request merged**